### PR TITLE
Battleship

### PR DESCRIPTION
**Pull request by albertlai431**
Soo battleship turned out to be a lil more complicated than I thought! But oh well, I guess it'll be the more advanced one

Notes:
- There are two versions, singleplayer and multiplayer. In singleplayer (VS the computer), the computer's strategy is simply random, though we could add difficulty levels in the future. 
- Also didn't add color to the terminal, was feeling sorta lazy lmao
- I used the sleep function to add some waiting times such that things don't just output like crazy really quickly, lmk if you think the wait times are too long though. 
- Would recommend looking at the multiplayer version first, a lot of the same code/concepts apply to the single player one too
- If you wanna check the rules out, take a look at the planning.md file

Hopefully the code is easy to understand but not guaranteeing anything 😅

Concepts Used:
- Data structures: 2D arrays, sets, tuples (no dictionary)
- Everything from lessons 1, 2 and 4

### CODE REVIEW

File: battleship/battleship-multi.py [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+            startCoords = convert(start)
+            if startCoords is None:
+                print(
+                    'Invalid input. Examples of proper coordinates: a0, B9, c2. Please try again.')
+                continue
+
+            # Get end coordinate and check validity
+            end = input('Coordinate for the end of your ' +
+                        shipNames[i] + '? ')
+            endCoords = convert(end)
```
Comment (EdZ543) [inline battleship/battleship-multi.py:37]: test

File: battleship/battleship-multi.py [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+    gameLength = 0
+    # Track coordinates for ships alive and sunk for both players in order to update char from 'o' to 'x' when the entire ship is sunk. Each index represents 1 ship.
+    playerOneShipAliveCoords = [set() for i in range(5)]
+    playerOneShipSunkCoords = [set() for i in range(5)]
+    playerTwoShipAliveCoords = [set() for i in range(5)]
+    playerTwoShipSunkCoords = [set() for i in range(5)]
+
+    print('Welcome to Battleship!')
+    print("--------------------------------------------------")
+    sleep(2)
```
Comment (edzhuang) [inline battleship/battleship-multi.py]: Pause is not really necessary IMO, there’s not much being printed

File: battleship/battleship-multi.py [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+    playerTwoGrid = [[' ' for i in range(10)] for j in range(10)]
+    # Keep track of ships alive for player1 and player2 to determine when to end the game
+    playerOneShipsAlive = 5
+    playerTwoShipsAlive = 5
+    gameLength = 0
+    # Track coordinates for ships alive and sunk for both players in order to update char from 'o' to 'x' when the entire ship is sunk. Each index represents 1 ship.
+    playerOneShipAliveCoords = [set() for i in range(5)]
+    playerOneShipSunkCoords = [set() for i in range(5)]
+    playerTwoShipAliveCoords = [set() for i in range(5)]
+    playerTwoShipSunkCoords = [set() for i in range(5)]
```
Comment (edzhuang) [inline battleship/battleship-multi.py:245]: I’m pretty sure we haven’t taught list comprehension, so either replace them or we can add list comprehension to lesson 3

File: battleship/battleship-multi.py [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+    for i in range(-1, 10):
+        for j in range(-1, 10):
+            # Print first row of letters
+            if i == -1:
+                if j == -1:
+                    print(" ", end=" ")
+                elif j == 9:
+                    print(chr(j+48))
+                else:
+                    print(chr(j+48), end=" ")
```
Comment (edzhuang) [inline battleship/battleship-multi.py]: chr(j+48) can just be replaced with j
```suggestion
                    print(j)
                else:
                    print(j, end=" ")
```

File: battleship/battleship-single.py [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+            startCoords = (randint(0, 9), randint(0, 9))
+            # Get random end coordinate from start coordinate to ensure proper ship length
+            endCoords = tuple(map(add, startCoords, choice(
+                [(0, shipLength[i]-1), (shipLength[i]-1, 0)])))
+            # Ensure end coordinate is on the grid
+            if endCoords[0] < 0 or endCoords[0] > 9 or endCoords[1] < 0 or endCoords[1] > 9:
+                continue
+            # Check validity of ship placement
+            if checkShipPlacement(startCoords, endCoords, i, computerShipAliveCoords):
+                print(str(startCoords) + str(endCoords))
```
Comment (edzhuang) [inline battleship/battleship-single.py]: This debugging line reveals the position of all of the computer’s ships, so probably don’t leave it in lol

**Review by edzhuang — CHANGES_REQUESTED** [review]
**General comments:**
- Very epic
- Definitely seems like the more advanced project, tbh I think it’d only be feasible for someone with prior programming experience
- The use of converting to unicode to print letters and convert coordinates was cool

**To fix:**
- Your attacks are shown again right after you sink a ship, maybe it’d be helpful for this to happen after a regular turn?
- Figuring out a correct end coordinate when placing ships can be tricky, maybe asking which direction the ship should extend out to would be easier

**Other comments:**
- The use of built-in functions like ord() and char() is fine because we introduced the concept of built-in functions, but I think a beginner would just use a list with the letters of the alphabet in them

File: battleship/battleship-multi.py (hunk 3 above)
Comment (albertlai431) [reply battleship/battleship-multi.py:245]: Hmmm I feel like list comprehension is pretty useful to learn but it's also a little complicated, maybe we can just touch a little bit on it during the lesson? We can ask the channel

File: battleship/battleship-multi.py (hunk 4 above)
Comment (albertlai431) [reply battleship/battleship-multi.py]: Oops yes that would be much better

File: battleship/battleship-single.py (hunk 5 above)
Comment (albertlai431) [reply battleship/battleship-single.py]: Oh CRAP I forgot to take that out LOL

**Comment by albertlai431** [comment]
@EdZ543 
> * Your attacks are shown again right after you sink a ship, maybe it’d be helpful for this to happen after a regular turn?

Yeah, I was going to do that originally but then things got really crowded, so I thought it'd just be simpler to only say if it's a hit or miss unless a ship was sunk. Do you think that works too?

File: battleship/battleship-multi.py (hunk 3 above)
Comment (albertlai431) [reply battleship/battleship-multi.py:245]: Okay I think we'll go with the latter option then!

**Comment by albertlai431** [comment]
Okay I finished changing almost everything I think! I changed the ship placements to direction rather than end coordinate, hopefully the logic for that is correct. That should be the main thing I think, the other changes were pretty small!

**Review by edzhuang — APPROVED** [review]
Aight the direction placing works goodly, I don't think there needs to be another reviewer
𝓨𝓸𝓾 𝓶𝓪𝔂 𝓱𝓮𝓻𝓮𝓫𝔂 𝓷𝓸𝔀 𝓶𝓮𝓻𝓰𝓮 𝓽𝓱𝓮 𝓫𝓻𝓪𝓷𝓬𝓱

**Pull request merged**